ionc/inc/ion_types.h

```diff
@@ -137,6 +137,7 @@ typedef struct _ion_collection          ION_COLLECTION;
 typedef struct _ion_stream        ION_STREAM;
 
 // decl's for user managed stream
+struct _ion_user_stream;
```

Comment:
So what we're doing here is declaring the struct, referencing the declaration, and then defining it - all within a few lines. This is the idiomatic (well, maybe just correct) way to handle these kinds of circular definitions.